lemur/plugins/lemur_acme/ultradns.py

```diff
@@ -0,0 +1,323 @@
+
+    for txt_record in txt_records:
+        if txt_record == token:
+            metrics.send("has_dns_propagated_success", "counter", 1)
+            return True
+
+    return False
+
+
+def wait_for_dns_change(change_id, account_number=None):
+    # Waits and checks if the DNS changes have propagated or not.
+    # First check the domains authoritative server. Once this succeeds,
+    # we ask a public DNS server (Google <8.8.8.8> in our case).
+    fqdn, token = change_id
+    number_of_attempts = 20
+    nameserver = get_authoritative_nameserver(fqdn)
+    for attempts in range(0, number_of_attempts):
+        status = _has_dns_propagated(fqdn, token, nameserver)
+        current_app.logger.debug("Record status on ultraDNS authoritative server for fqdn: {}: {}".format(fqdn, status))
+        if status:
+            time.sleep(10)
+            break
+        time.sleep(10)
+    if status:
+        for attempts in range(0, number_of_attempts):
+            status = _has_dns_propagated(fqdn, token, get_public_authoritative_nameserver())
+            current_app.logger.debug("Record status on Google DNS for fqdn: {}: {}".format(fqdn, status))
+            if status:
+                metrics.send("wait_for_dns_change_success", "counter", 1)
+                break
+            time.sleep(10)
+    if not status:
+        metrics.send("wait_for_dns_change_fail", "counter", 1)
+        sentry.captureException(extra={"fqdn": str(fqdn), "txt_record": str(token)})
+        metrics.send(
+            "wait_for_dns_change_error",
+            "counter",
+            1,
+            metric_tags={"fqdn": fqdn, "txt_record": token},
+        )
+    return
+
+
+def get_zones(account_number):
+    # Get zones from the UltraDNS
+    path = "/v2/zones"
+    zones = []
+    for page in _paginate(path, "zones"):
+        for elem in page:
+            # UltraDNS zone names end with a "." - Example - lemur.example.com.
+            # We pick out the names minus the "." at the end while returning the list
+            zone = Zone(elem)
+            if zone.authoritative_type == "PRIMARY" and zone.status == "ACTIVE":
+                zones.append(zone.name)
+
+    return zones
+
+
+def get_zone_name(domain, account_number):
+    # Get the matching zone for the given domain
+    zones = get_zones(account_number)
+    zone_name = ""
+    for z in zones:
+        if domain.endswith(z):
+            # Find the most specific zone possible for the domain
+            # Ex: If fqdn is a.b.c.com, there is a zone for c.com,
+            # and a zone for b.c.com, we want to use b.c.com.
+            if z.count(".") > zone_name.count("."):
+                zone_name = z
+    if not zone_name:
+        metrics.send("ultradns_no_zone_name", "counter", 1)
+        raise Exception("No UltraDNS zone found for domain: {}".format(domain))
+    return zone_name
+
+
+def create_txt_record(domain, token, account_number):
+    # Create a TXT record for the given domain.
```

Comment:
Convert to docstring? Recommend for all of your functions: https://www.python.org/dev/peps/pep-0257/